### ISSUE TITLE
Refactor the process of calculating evaluation metrics

### ISSUE THREAD

**Issue by LiMinghao1994** [sql-machine-learning/elasticdl#1082]
# Currently Existing Problems
Currently, during the process of evaluation, workers report metrics calculated from every minibatch to master. We assume that the metrics reported in every minibatch was independent and master calculates the average of those metrics.

However, metrics inherited from `tf.keras.metrics` does not behave like this, e.g. `tf.keras.metrics.AUC`. These are streaming metrics, i.e. each time they are called, they calculate the metrics from all the minibatches they have received so far. More specifically, when `tf.keras.metrics.AUC` receives the model outputs and labels of the second batch, it calculates AUC of the first two batches rather than the second batch only. 

As a result, in order to support metrics inherited from `tf.keras.metrics`, we should refactor the process of calculating evaluation metrics.

# New Evaluation Metrics Interface
Referring to the evaluation metric interface of Keras, we plan to support two kinds of interfaces for calculating evaluation metrics:
* Keras-native streaming metrics, i.e. metric classes in `tf.keras.metrics`.
* Stateless functions will also be supported. However, we require the arguments of the stateless functions to be `(lables, model_outputs)`, just like [Keras](https://keras.io/metrics/#custom-metrics) has required. 

An example of new interfaces:
```
def eval_metrics_fn():
    return {
        "acc": tensorflow.keras.metrics.Accuracy(),
        "mse": tensorflow.keras.metrics.MeanSquaredError(),
        "acc_fn": lambda lables, outputs: tf.equal(
            tf.cast(labels, tf.int32), 
            tf.cast(outputs, tf.int32), 
        ),
    }
```

# How to calculate evaluation metrics
For metric instances of `tf.keras.metrics.Metric`, the logic of calculating evaluation metrics can be represented by the following code (take AUC for an example):
```
metric_instance = tf.keras.metrics.AUC()
for minibatch_data in dataset:
    inputs, labels = minibatch_data
    outputs = forward_pass(inputs)
    metric_instance.update_state(labels, outputs)
print(metric_instance.result())
```

For stateless functions, Keras provides `MeanMetricWrapper` which wraps a stateless function into a streaming metric instance.

Please note that evaluation process may involve multiple workers, each worker calculates a subset of evaluation data. Thus, workers should report model outputs and labels to the master, and the master call `metric_instance.update_state`.

# Different Metrics for Different Model Outputs
Some models may have multiple outputs. According to compile method part in  Keras [official tutorial](http://faroit.com/keras-docs/2.0.4/models/model/), Keras supports users to specify different metrics for different model outputs, such as `metrics={'output_a': 'accuracy', 'output_b': 'AUC'}`. This is necessary because calculating AUC requires that model has multiple outputs.
ElasticDL support models with multiple outputs now. We should also support different metrics for different outputs. Thus we provide two kinds of evaluation metric interfaces:
```
# Interface for models with only one output. 
# ElasticDL will calculate "acc" and "mse" using the model output.
def eval_metrics_fn_for_one_output():
    return {
        "acc": tensorflow.keras.metrics.Accuracy(),
        "mse": tensorflow.keras.metrics.MeanSquaredError(),
    }

# Interface for models with multiple outputs. 
# For model output "output_a", ElasticDL will calculate "acc" and "mse". 
# For "output_b", ElasticDL will calculate "auc".
def eval_metrics_fn_for_multiple_outputs():
    return {
        "output_a": {
            "acc": tensorflow.keras.metrics.Accuracy(),
            "mse": tensorflow.keras.metrics.MeanSquaredError(),
        },
        "output_b": {
            "auc": tensorflow.keras.metrics.AUC(),
        },
    }
```

**Comment by LiMinghao1994** [comment]
After discussing with @workingloong and @skydoorkai and checking the code of `tf.keras.metrics`, we find that there is no risk of OOM. So there is no need to start a new pod. Workers report model outputs and labels to the master, the master calculates metrics.

**Comment by LiMinghao1994** [comment]
# New Evaluation Metrics Interface
Referring to the evaluation metric interface of Keras, I plan to support two kinds of interfaces for calculating evaluation metrics:
* Keras-native streaming metrics, i.e. metric classes in `tf.keras.metrics`.
* Stateless functions will also be supported. However, we require the arguments of the stateless functions to be `(lables, model_outputs)`, just like [Keras](https://keras.io/metrics/#custom-metrics) has required. 

An example of new interfaces:
```
def eval_metrics_fn():
    return {
        "acc": tensorflow.keras.metrics.Accuracy(),
        "mse": tensorflow.keras.metrics.MeanSquaredError(),
        "acc_fn": lambda lables, outputs: tf.equal(
            tf.cast(labels, tf.int32), 
            tf.cast(outputs, tf.int32), 
        ),
    }
```

# How to calculate evaluation metrics
For metric instances of `tf.keras.metrics.Metric`, the logic of calculating evaluation metrics can be represented by the following code (take AUC for an example):
```
metric_instance = tf.keras.metrics.AUC()
for minibatch_data in dataset:
    inputs, labels = minibatch_data
    outputs = forward_pass(inputs)
    metric_instance.update_state(labels, outputs)
print(metric_instance.result())
```

For stateless functions, Keras provides `MeanMetricWrapper` which wraps a stateless function into a streaming metric instance.

Please note that evaluation process may involve multiple workers, each worker calculates a subset of evaluation data. Thus, workers should report model outputs and labels to the master, and the master call `metric_instance.update_state`.
